src/opnsense/mvc/app/controllers/OPNsense/Core/Api/BootEnvironmentController.php

```diff
@@ -0,0 +1,243 @@
+     */
+    public function isSupportedAction()
+    {
+        $result = json_decode((new Backend())->configdRun('bootenvironment supported'), true) ?? [];
+        return ['supported' => !empty($result) && $result['status'] == 'OK'];
+    }
+
+    /**
+     * search boot environments
+     * @return array
+     */
+    public function searchAction()
+    {
+        $records = json_decode((new Backend())->configdRun('bootenvironment list'), true) ?? [];
+        return $this->searchRecordsetBase($records);
+    }
+
+    /**
+     * fetch an environment by uuid, return new when not found or $uuid equals null
+     * @param string $uuid
+     * @return array
+     */
+    public function getAction($uuid = null)
+    {
+        if (!empty($uuid)) {
+            $result = $this->findByUuid($uuid);
+            if (!empty($result)) {
+                return $result;
+            }
+        }
+        // new or not found
+        return ['name' => 'BE'.date("YmdHis"), 'uuid' => ''];
+    }
+
+    /**
+     * create a new bootenvironment
+     * @param string $uuid uuid to save
+     * @return array status
+     */
+    public function setAction($uuid)
+    {
+        if ($this->request->isPost() && $this->request->hasPost('name')) {
+            $name =  $this->request->getPost('name', 'string', null);
+
+            $be = $this->findByUuid($uuid);
+            $new_be = $this->findByName($name);
+
+            if (!empty($be) && $be['name'] == $name) {
+                /* skip, unchanged */
+                return ['status' => 'ok'];
+            } elseif (!empty($be) && empty($new_be) && $this->isValidName($name)) {
+                return json_decode(
+                    (new Backend())->configdpRun("bootenvironment rename", [$be['name'], $name]),
+                    true
+                );
+            } else {
+                if (!empty($new_be)) {
+                    $msg = gettext('A boot environment already exists by this name');
+                } elseif (!$this->isValidName($name)){
+                    $msg = gettext('Invalid name specified');
+                } else {
+                    $msg = gettext('Boot environment not found');
+                }
+                return [
+                    'status' => 'failed',
+                    'validations' => [
+                        'name' => $msg
+                    ]
+                ];
+            }
+        }
+
+        return ['status' => 'failed'];
+    }
+
+    /**
+     * add or clone a boot environment
+     * @return array status
+     */
+    public function addAction()
+    {
+        if ($this->request->isPost()) {
+            $uuid =  $this->request->getPost('uuid', 'string', '');
+            $name =  $this->request->getPost('name', 'string', '');
+
+            $msg = null;
+            if ($this->findByName($name)) {
+                $msg = gettext('A boot environment already exists by this name');
```

Comment:
we may want to consider to rename all the "boot environment" strings to "snapshot" too. Similar to "BE" which is highly ambiguous in our scope.